### PR TITLE
SDK-992 implement task manifest generator

### PR DESCRIPTION
**Pull request by sergeyspryker**
https://spryker.atlassian.net/browse/SDK-992

### CODE REVIEW

File: src/Infrastructure/Manifest/Interaction/ManifestInteractionProcessor.php [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+                return $this->setValueByKeyStack($receivedValues, $keyStack, $valueConfig->getValue());
+            case $valueConfig instanceof ValueCollection:
+                return $this->processCollection($id, $valueConfig, $keyStack, $receivedValues);
+            case $valueConfig instanceof CallbackValue:
+                return $this->processCallbackValue($id, $valueConfig, $keyStack, $receivedValues);
+            case is_array($valueConfig):
+                return $this->processNested($valueConfig, $keyStack, $receivedValues);
+            default:
+                throw new InvalidArgumentException(sprintf('Invalid mapping in `%s` node', $id));
+        }
```
Comment (sergeyspryker) [inline src/Infrastructure/Manifest/Interaction/ManifestInteractionProcessor.php:96]: We don't need polymorphism there due to small class and recursion

File: extension/Custom/src/DependencyInjection/CustomExtension.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace Custom\DependencyInjection;
+
+use Symfony\Component\Config\FileLocator;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Extension\Extension;
+use Symfony\Component\DependencyInjection\Extension\PrependExtensionInterface;
+use Symfony\Component\DependencyInjection\Loader\YamlFileLoader;
+
+class CustomExtension extends Extension implements PrependExtensionInterface
```
Comment (vol4onok) [inline extension/Custom/src/DependencyInjection/CustomExtension.php:16]: Why do we need this extension?

File: src/Core/Domain/Enum/TaskTypeEnum.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+/**
+ * Copyright © 2019-present Spryker Systems GmbH. All rights reserved.
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace SprykerSdk\Sdk\Core\Domain\Enum;
+
+final class TaskTypeEnum
```
Comment (vol4onok) [inline src/Core/Domain/Enum/TaskTypeEnum.php:10]: it should be in the Contracts. and it's already exist in another PR

File: src/Infrastructure/Manifest/Interaction/ManifestInteractionProcessor.php [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+     * @param \SplStack $keyStack
+     * @param mixed $value
+     *
+     * @return array
+     */
+    protected function setValueByKeyStack(array $receivedValues, SplStack $keyStack, $value): array
+    {
+        $where = &$receivedValues;
+
+        foreach (array_reverse(iterator_to_array($keyStack)) as $key) {
```
Comment (vol4onok) [inline src/Infrastructure/Manifest/Interaction/ManifestInteractionProcessor.php:192]: what's the magic with links? pls rewrite it

**Review by vol4onok — CHANGES_REQUESTED** [review]

**Comment by vol4onok** [comment]
Pls merge last changes from master

File: src/Infrastructure/Manifest/Interaction/ManifestInteractionProcessor.php (hunk 4 above)
Comment (sergeyspryker) [reply src/Infrastructure/Manifest/Interaction/ManifestInteractionProcessor.php:192]: Not the magic. This is most optimal way to populate array by sequence of keys placed in another array

File: src/Infrastructure/Manifest/Interaction/ManifestInteractionProcessor.php [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+     *
+     * @return array
+     */
+    protected function processCollection(
+        string $id,
+        ValueCollection $valuesCollection,
+        SplStack $keyStack,
+        array $receivedValues
+    ): array {
+        $index = 0;
```
[the remark this answers is not in the document]
Comment (sergeyspryker) [reply src/Infrastructure/Manifest/Interaction/ManifestInteractionProcessor.php:133]: it isn't shown in console

File: extension/Custom/src/DependencyInjection/CustomExtension.php (hunk 2 above)
Comment (sergeyspryker) [reply extension/Custom/src/DependencyInjection/CustomExtension.php:16]: This is a custom bundle for placing custom generated Task/TakSets and etc.

**Review by vol4onok — APPROVED** [review]

File: src/Infrastructure/Manifest/ManifestWriter/Task/FormatManifestWriter.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace SprykerSdk\Sdk\Infrastructure\Manifest\ManifestWriter\Task;
+
+use SprykerSdk\Sdk\Core\Application\Dto\Manifest\ManifestFile;
+use SprykerSdk\Sdk\Infrastructure\Manifest\ManifestWriter\ManifestFormatWriterInterface;
+use Symfony\Component\Filesystem\Filesystem;
+
+abstract class FormatManifestWriter implements ManifestFormatWriterInterface
```
Comment (pavelmaksimov25) [inline src/Infrastructure/Manifest/ManifestWriter/Task/FormatManifestWriter.php:14]: what the point having it abstract?

File: src/Infrastructure/Manifest/TemplateReader/Task/FormatTaskTemplateReader.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace SprykerSdk\Sdk\Infrastructure\Manifest\TemplateReader\Task;
+
+use SprykerSdk\Sdk\Core\Application\Dto\Manifest\ManifestRequestDtoInterface;
+use SprykerSdk\Sdk\Infrastructure\Manifest\TemplateReader\TemplateFormatReaderInterface;
+use Twig\Environment;
+
+abstract class FormatTaskTemplateReader implements TemplateFormatReaderInterface
```
Comment (pavelmaksimov25) [inline src/Infrastructure/Manifest/TemplateReader/Task/FormatTaskTemplateReader.php:14]: What is the purpose of having this abstract?

File: src/Infrastructure/Resources/config/services.yaml [hunk 8]
```diff
@@ -567,7 +573,7 @@ services:
             - "@task.task_set_placeholders_builder"
             - "@task.task_set_commands_builder"
 
-    # Validators
+    # Validators1
```
Comment (pavelmaksimov25) [inline src/Infrastructure/Resources/config/services.yaml:576]: typo

File: src/Infrastructure/Resources/config/services.yaml [hunk 9]
```diff
@@ -584,6 +590,65 @@ services:
...
+            - "@cli_interaction_processor"
+
+    manifest.interaction.processor:
+        class: SprykerSdk\Sdk\Infrastructure\Manifest\Interaction\ManifestInteractionProcessor
+        arguments:
+            - "@manifest.interaction.need_to_ask_question"
+            - "@manifest.interaction.value_question"
+            - "@manifest.interaction.new_collection_item_question"
+
+    manifest.task_php_template_format_reader:
```
Comment (pavelmaksimov25) [inline src/Infrastructure/Resources/config/services.yaml:615]: `manifest.reader.task_php_template`. The same pattern for the rest of the readers and writers

File: tests/Sdk/Unit/Core/Application/Manifest/Task/TaskManifestGeneratorTest.php [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        // Assert
+        $this->assertSame('test/Task.yaml', $response->getCreatedFileName());
+    }
+
+    /**
+     * @param string $filePath
+     *
+     * @return \SprykerSdk\Sdk\Core\Application\Dependency\Manifest\ManifestWriterInterface
+     */
+    public function createManifestWriterMock(string $filePath): ManifestWriterInterface
```
Comment (pavelmaksimov25) [inline tests/Sdk/Unit/Core/Application/Manifest/Task/TaskManifestGeneratorTest.php:54]: can be changed to protected, 'cause it is not a data provider

File: src/Presentation/Console/Command/GenerateTaskCommand.php [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+     */
+    protected function updateTask(OutputInterface $output): void
+    {
+        $output->writeln('');
+        $output->writeln('Updating tasks...');
+
+        $application = $this->getApplication();
+
+        if ($application === null) {
+            throw new RuntimeException('Ban\'t find current console application');
```
Comment (pavelmaksimov25) [inline src/Presentation/Console/Command/GenerateTaskCommand.php:138]: `Can't`

File: src/Presentation/Console/Command/GenerateTaskCommand.php [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        $commandExecutionCode = $application
+            ->find(UpdateSdkCommand::NAME)
+            ->run(new ArrayInput([]), $output);
+
+        if ($commandExecutionCode !== static::SUCCESS) {
+            throw new RuntimeException('Task update failed. Try to update it manually');
```
Comment (pavelmaksimov25) [inline src/Presentation/Console/Command/GenerateTaskCommand.php:146]: I think we need to provide original error as well

**Review by pavelmaksimov25 — CHANGES_REQUESTED** [review]

File: src/Infrastructure/Manifest/Interaction/ManifestInteractionProcessor.php (hunk 1 above)
Comment (pavelmaksimov25) [reply src/Infrastructure/Manifest/Interaction/ManifestInteractionProcessor.php:96]: Unfortunately, I must insist on polymorphic approach

File: src/Infrastructure/Manifest/Interaction/Question/NeedToAskQuestion.php [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        $this->interactionProcessor = $interactionProcessor;
+    }
+
+    /**
+     * @param string $valueId
+     *
+     * @return bool
+     */
+    public function ask(string $valueId): bool
```
[the remark this answers is not in the document]
Comment (sergeyspryker) [reply src/Infrastructure/Manifest/Interaction/Question/NeedToAskQuestion.php:34]: Discussed. We unable to unify input/output types

File: src/Infrastructure/Validator/Manifest/TaskManifestConfiguration.php [hunk 14]
```diff
@@ -63,7 +63,7 @@ public function getConfigTreeBuilder(array $config): TreeBuilder
                         ->thenInvalid('Task id `%s` should have `/^[a-z-]+:[a-z-]+:[a-z-])+$/` format.')
                     ->end()
                 ->end()
-                ->scalarNode('stage')->isRequired()->end()
+                ->scalarNode('stage')->end()
```
[the remark this answers is not in the document]
Comment (sergeyspryker) [reply src/Infrastructure/Validator/Manifest/TaskManifestConfiguration.php]: Doesn't work like that
`->scalarNode('stage')->isRequired()->defaultValue(ContextInterface::DEFAULT_STAGE)->end()`
only
`->scalarNode('stage')->defaultValue(ContextInterface::DEFAULT_STAGE)->end()`

**Review by pavelmaksimov25 — APPROVED** [review]

**Pull request merged**